File: temas/04-kmedoids/codigo.R
# ==========================================================================
# TEMA 04: K-MEDOIDES
# Clase 3
# ==========================================================================


# --------------------------------------------------------------------------
# PAM y distancia de Gower en R
# Origen: Equivalente agregado, no viene del material del curso. La función daisy con metric gower si aparece en el código de MDS del curso
# --------------------------------------------------------------------------

library(cluster)

# Caso euclidiano
Xs <- scale(iris[, 1:4])
pam_res <- pam(Xs, k = 3)

pam_res$medoids     # coordenadas de los medoides
pam_res$id.med      # índices de los medoides dentro de la muestra
pam_res$clustering  # etiquetas
pam_res$objective   # valor de la función objetivo

plot(Xs[, 3], Xs[, 4], col = pam_res$clustering, pch = 16)
points(pam_res$medoids[, 3], pam_res$medoids[, 4], pch = 8, cex = 2)

# Silhouette de la solución
plot(silhouette(pam_res))


# Caso no euclidiano: datos mixtos con distancia de Gower
xx <- read.table("sponge.txt", header = TRUE, sep = ",")
dis_gower <- daisy(xx, metric = "gower")

pam_gower <- pam(dis_gower, k = 3, diss = TRUE)
pam_gower$id.med
pam_gower$clustering


# CLARA: versión de PAM para muestras grandes
clara_res <- clara(Xs, k = 3, samples = 50)
clara_res$medoids

File: temas/04-kmedoids/codigo.py
# ==========================================================================
# TEMA 04: K-MEDOIDES
# Clase 3
# ==========================================================================


# --------------------------------------------------------------------------
# K-Medoides con distancia euclidiana
# Origen: Del curso, class3_book1_kmedoids.ipynb
# --------------------------------------------------------------------------

from scipy.spatial import distance_matrix, distance
import numpy as np
import matplotlib.pyplot as plt

# La entrada del algoritmo es la matriz de distancias n x n
dista_s = distance_matrix(Xs, Xs)
np.shape(dista_s)

#%pip install kmedoids   # quitar el # si falta instalar
import kmedoids

c = kmedoids.fasterpam(dista_s, 3)
print("Loss is:", c.loss)

labels = c.labels      # etiquetas de grupo
medoids = c.medoids    # ÍNDICES de las observaciones medoide

plt.scatter(Xs[:, 2], Xs[:, 3], c=labels, cmap='viridis', alpha=0.6)
plt.title("K-Medoids Clustering")
plt.show()


# --------------------------------------------------------------------------
# K-Medoides con distancia de Gower (datos no euclidianos)
# Origen: Del curso, class3_book1_kmedoids.ipynb
# --------------------------------------------------------------------------

#!pip install gower   # quitar el # si falta instalar
import gower

from sklearn.datasets import load_wine
wine = load_wine()
W = wine.data

# Matriz de distancias tipo Gower: sirve para variables mixtas o categóricas
dist_g = gower.gower_matrix(W)
np.shape(dist_g)

med_w = kmedoids.fasterpam(dist_g, 3)

labels = med_w.labels
medoids = med_w.medoids

print(medoids)   # índices de los medoides
print(labels)


# --------------------------------------------------------------------------
# Alternativa con sklearn_extra
# Origen: Equivalente agregado, no viene del material del curso
# --------------------------------------------------------------------------

# Si kmedoids no está disponible
from sklearn_extra.cluster import KMedoids

modelo = KMedoids(n_clusters=3, metric='euclidean', method='pam', random_state=0)
etiquetas = modelo.fit_predict(Xs)

modelo.medoid_indices_    # índices de los medoides
modelo.cluster_centers_   # coordenadas de los medoides
modelo.inertia_           # suma de distancias a los medoides

# Con una matriz de distancias precalculada
modelo_pre = KMedoids(n_clusters=3, metric='precomputed', method='pam', random_state=0)
etiquetas_pre = modelo_pre.fit_predict(dista_s)
